### PR TITLE
[Proposal] Default classes to be non-subclassable publicly

### PR DESCRIPTION
**Pull request by JaviSoto**
Mailing list discussion can be found [here](http://thread.gmane.org/gmane.comp.lang.swift.evolution/9702/focus=9708)

### CODE REVIEW

File: proposals/0000-sealed-by-default.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+possibility.
+- Requiring that the author of a class mark a class as `open` is akin to requiring
+symbols to be explicitly `public`: it ensures that a conscious decision is made
+regarding whether the ability to subclass a `class` is part of the API.
+
+## Proposed solution
+
+- New `sealed` class modifier for classes and methods which marks them as only
+overridable within the module they're declared in.
+- `sealed` because the default for classes and methods.
```
Comment (neonichu) [inline proposals/0000-sealed-by-default.md]: > becomes

Comment (JaviSoto) [inline proposals/0000-sealed-by-default.md]: Fixed, thanks!

**Comment by xnukernel** [comment]
I'm not sure I would agree with the premise here. In Swift `struct` is intentionally made more powerful, thus the clear distinction between a structure and a class is inheritability. A struct may still conform to protocols that ensure compliant behavior but may not inherit behavior nor is it inheritable. I think this an important distinction you're trying to achieve with your keyword. Your behavior is of `sealed` is exactly what the `convenience` keyword accomplishes, no? Am I missing something? If anything I would propose a change to the keyword `convenience` to be called something along the lines of `sealed`.

**Comment by JaviSoto** [comment]
> thus the clear distinction between a structure and a class is inheritability

That's only one of the distinctions. I would argue that the "value" vs "identity" semantics is a more important difference.

> Your behavior is of sealed is exactly what the convenience keyword accomplishes, no? 

I'm not sure I follow this. `convenience` applies specifically to constructors.

Something important about this proposal that maybe I should note in it is the fact that this doesn't really introduce any new concept to the language: `final` already exists, and allows you to declare classes that don't allow for inheritance. The goal of this proposal is to aid with API design by having a more sensible default, the same way `internal` is the default for visibly.

**Comment by lattner** [comment]
@rjmccall Please take a look at this.  One random comment, "sealed" is not a term we use in swift.  Please use the term "not subclassable", because "sealed" leaves a lot open to interpretation.

**Comment by lattner** [comment]
Ah, please disregard my comment about "sealed".  I thought that this was a proposal to change the default behavior of public classes to not be subclassable.

**Comment by JaviSoto** [comment]
I'm confused @lattner, that is what this proposal is about :-) (and I'm happy to change the specific name of the keyword itself)

**Comment by lattner** [comment]
Yeah, I'm probably just totally confused.  Please work with John on this and email me off list when it is ready for merging + review discussion.

**Comment by JaviSoto** [comment]
Will do! Thanks, Chris.

**Comment by xnukernel** [comment]
@JaviSoto okay I see what you're saying. I was thoroughly confused earlier. 

I'm not 100% convinced that the default behavior you're suggesting is common as you're suggesting, enough to force the user to specify heritable traits (methods/properties). I think this breaks down the entire inheritance as a key concept of OO programming. I'd be worried of code riddled with this new keyword for behavior one would expect from an OO programming language.

I would agree that `sealed` is probably an ambiguous keyword, it should be apparent what the behavior they keyword is applying, I'm not sure `sealed` is a metaphor that is intuitive. The `heritable` keyword makes more sense, but is a pretty ugly word to have in Swift syntax.

**Comment by rjmccall** [comment]
@xnukernel It's more that we don't want the default behavior to make implicit promises that limit the future evolution of a library.  Designing a class properly for subclassing is actually quite difficult; any sort of delegation to an overridable operation has to be carefully considered and documented, and there has to be an understanding of how any new operations will interact with the existing ones.  Merely making a class a public part of an API should not promise that these things are settled.

That said, I share some of your concerns about this creating a lot of extra annotations.  Since it's only meaningful on public classes/methods anyway, maybe we can consider "overridable" to be an extra tier of access control?  That would imply that the name should include or strongly imply "public".

One disadvantage of that, though, is that ordinary "public" would be equivalent to "sealed", so there wouldn't be a way to explicit about your intent to make something public and not overridable.

**Comment by lattner** [comment]
@JaviSoto and @rjmccall when this is ready to go, please email me.  I'd like to kick it off on Tuesday if possible.

**Comment by JaviSoto** [comment]
@lattner I updated it with @rjmccall's feedback. I emailed John to get his opinion. I'll update it with his feedback and let you know, from my end this is ready to go for Tuesday :)

**Comment by rjmccall** [comment]
The revision looks great to me, thanks!  I think it's ready to go.

**Pull request merged**